esp/services/ws_workunits/ws_workunitsQuerySets.cpp

```diff
@@ -2859,6 +2887,159 @@ bool CWsWorkunitsEx::onWUQuerysetCopyQuery(IEspContext &context, IEspWUQuerySetC
     return true;
 }
 
+bool CWsWorkunitsEx::onWUQuerysetImport(IEspContext &context, IEspWUQuerysetImportRequest &req, IEspWUQuerysetImportResponse &resp)
+{
+    try
+    {
+        const char* target = req.getTarget();
+        if (!target || !*target)
+            throw MakeStringException(ECLWATCH_QUERYSET_NOT_FOUND, "Target not specified");
+
+        MemoryBuffer &mb = const_cast<MemoryBuffer &>(req.getData()); //for efficiency, content of request shouldn't matter after
+        if (req.getCompressed())
+        {
+            MemoryBuffer decompressed;
+            fastLZDecompressToBuffer(decompressed, mb);
+            mb.swapWith(decompressed);
+        }
+        mb.append('\0');
+
+        Owned<IPropertyTree> srcTree = createPTreeFromXMLString(mb.toByteArray());
+        const char *archivedTarget = srcTree->queryProp("@target");
+        if (archivedTarget && *archivedTarget) //support simple queryset or with archived (exported) root format
+        {
+            VStringBuffer xpath("QuerySet[@id='%s']", archivedTarget);
+            IPropertyTree *qsTree = srcTree->queryPropTree(xpath);
+            if (qsTree)
+                srcTree.setown(LINK(qsTree));
+        }
+        if (req.getReplace())
+        {
+            Owned<IPropertyTree> queryRegistry = getQueryRegistry(target, false);
+            queryRegistry->removeProp("*");
+            resp.setClearedExisting(true);
+        }
+
+        const bool activate = CQuerysetImportActivation_ImportedActive == req.getActivation(); //only two options now but may evolve
+
+        QueryCloner cloner(&context, srcTree, target);
+
+        SCMStringBuffer process;
+        if (req.getCopyFiles())
+        {
+            Owned <IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(target);
+            if (clusterInfo && clusterInfo->getPlatform()==RoxieCluster)
+            {
+                clusterInfo->getRoxieProcess(process);
+                if (!process.length())
+                    throw MakeStringException(ECLWATCH_INVALID_CLUSTER_INFO, "DFS process cluster not found for destination target %s", target);
+                unsigned updateFlags = 0;
+                if (req.getOverwriteDfs())
+                    updateFlags |= (DALI_UPDATEF_REPLACE_FILE | DALI_UPDATEF_CLONE_FROM | DALI_UPDATEF_SUPERFILES);
+                if (req.getUpdateCloneFrom())
+                    updateFlags |= DALI_UPDATEF_CLONE_FROM;
+                if (req.getUpdateSuperFiles())
+                    updateFlags |= DALI_UPDATEF_SUPERFILES;
+                if (req.getAppendCluster())
+                    updateFlags |= DALI_UPDATEF_APPEND_CLUSTER;
+
+                cloner.enableFileCloning(updateFlags, req.getDfsServer(), process.str(), req.getSourceProcess(), req.getAllowForeignFiles());
+            }
```

Comment:
Should we throw an exception if the target is not a roxie?

Reply:
I'll move some code around and validate at the top of the method